File: js/fndsa/src/fft.js
/**
 * Complex FFT for FN-DSA (FIPS 206 / FALCON).
 * Negacyclic complex FFT over C[x]/(x^n+1).
 * Mirrors Python fft.py exactly.
 */

function fftLogN(n) {
  let logn = 0, t = n;
  while (t > 1) { t >>= 1; logn++; }
  return logn;
}

function fftBitRev(k, logn) {
  let r = 0;
  for (let i = 0; i < logn; i++) {
    r = (r << 1) | (k & 1);
    k >>= 1;
  }
  return r;
}

/**
 * Forward negacyclic complex FFT over C[x]/(x^n+1).
 * f: array of n numbers (real polynomial coefficients).
 * Returns array of n [re, im] pairs.
 */
export function fft(f, n) {
  const logn = fftLogN(n);
  // Copy input as complex pairs [re, im]
  const a = new Array(n);
  for (let i = 0; i < n; i++) a[i] = [f[i], 0.0];

  let k = 0;
  let length = n >> 1;
  while (length >= 1) {
    for (let start = 0; start < n; start += 2 * length) {
      k++;
      const brk = fftBitRev(k, logn);
      const angle = Math.PI * brk / n;
      const wRe = Math.cos(angle), wIm = Math.sin(angle);
      for (let j = start; j < start + length; j++) {
        const [aRe, aIm] = a[j];
        const [bRe, bIm] = a[j + length];
        // t = w * a[j+length]
        const tRe = wRe * bRe - wIm * bIm;
        const tIm = wRe * bIm + wIm * bRe;
        a[j + length] = [aRe - tRe, aIm - tIm];
        a[j] = [aRe + tRe, aIm + tIm];
      }
    }
    length >>= 1;
  }
  return a;
}

/**
 * Inverse negacyclic complex FFT over C[x]/(x^n+1).
 * f: array of n [re, im] pairs.
 * Returns array of n [re, im] pairs (scaled by 1/n).
 */
export function ifft(f, n) {
  const logn = fftLogN(n);
  const a = f.map(v => [v[0], v[1]]);

  let k = n;
  let length = 1;
  while (length < n) {
    for (let start = n - 2 * length; start >= 0; start -= 2 * length) {
      k--;
      const brk = fftBitRev(k, logn);
      const angle = Math.PI * brk / n;
      const wRe = Math.cos(angle), wIm = -Math.sin(angle); // conjugate = inverse twiddle
      for (let j = start; j < start + length; j++) {
        const [tRe, tIm] = a[j];
        const [bRe, bIm] = a[j + length];
        a[j] = [tRe + bRe, tIm + bIm];
        // a[j+length] = w_inv * (t - a[j+length])
        const dRe = tRe - bRe, dIm = tIm - bIm;
        a[j + length] = [wRe * dRe - wIm * dIm, wRe * dIm + wIm * dRe];
      }
    }
    length <<= 1;
  }
  const invN = 1.0 / n;
  return a.map(v => [v[0] * invN, v[1] * invN]);
}

/**
 * Split FFT-domain polynomial into two halves.
 * Given f in FFT domain (bit-reversed), computes f0, f1 where
 * f(x) = f0(x^2) + x*f1(x^2).
 */
export function splitFft(F, n) {
  const logn = fftLogN(n);
  const h = n >> 1;
  const f0 = new Array(h);
  const f1 = new Array(h);
  for (let k = 0; k < h; k++) {
    const j = fftBitRev(k, logn - 1);
    const angle = Math.PI * (2 * j + 1) / n;
    const omRe = Math.cos(angle), omIm = Math.sin(angle);
    const [aRe, aIm] = F[2 * k];
    const [bRe, bIm] = F[2 * k + 1];
    f0[k] = [(aRe + bRe) / 2, (aIm + bIm) / 2];
    // f1[k] = (a - b) / (2 * omega_j)
    const dRe = (aRe - bRe) / 2, dIm = (aIm - bIm) / 2;
    // Divide by omega_j: multiply by conj(omega_j) / |omega_j|^2, but |omega_j|=1
    // So divide by omega_j = multiply by conj(omega_j)
    f1[k] = [dRe * omRe + dIm * omIm, dIm * omRe - dRe * omIm];
  }
  return [f0, f1];
}

/**
 * Merge two FFT-domain half-polynomials into one.
 * Inverse of splitFft.
 */
export function mergeFft(f0, f1, n) {
  const logn = fftLogN(n);
  const h = n >> 1;
  const result = new Array(n);
  for (let k = 0; k < h; k++) {
    const j = fftBitRev(k, logn - 1);
    const angle = Math.PI * (2 * j + 1) / n;
    const omRe = Math.cos(angle), omIm = Math.sin(angle);
    const [f0Re, f0Im] = f0[k];
    const [f1Re, f1Im] = f1[k];
    // t = omega_j * f1[k]
    const tRe = omRe * f1Re - omIm * f1Im;
    const tIm = omRe * f1Im + omIm * f1Re;
    result[2 * k] = [f0Re + tRe, f0Im + tIm];
    result[2 * k + 1] = [f0Re - tRe, f0Im - tIm];
  }
  return result;
}

File: rust/fn-dsa/src/fft.rs
// Complex FFT/IFFT/SplitFFT/MergeFFT for FN-DSA (FIPS 206 / FALCON).
//
// Uses Rust f64 (IEEE 754 double), matching Go float64.

use std::f64::consts::PI;

pub type Complex64 = (f64, f64);

#[inline]
fn cadd(a: Complex64, b: Complex64) -> Complex64 {
    (a.0 + b.0, a.1 + b.1)
}

#[inline]
fn csub(a: Complex64, b: Complex64) -> Complex64 {
    (a.0 - b.0, a.1 - b.1)
}

#[inline]
fn cmul(a: Complex64, b: Complex64) -> Complex64 {
    (a.0 * b.0 - a.1 * b.1, a.0 * b.1 + a.1 * b.0)
}

#[inline]
fn cconj(a: Complex64) -> Complex64 {
    (a.0, -a.1)
}

#[inline]
fn cdiv_real(a: Complex64, r: f64) -> Complex64 {
    (a.0 / r, a.1 / r)
}

fn bit_rev(mut k: usize, logn: usize) -> usize {
    let mut r = 0usize;
    for _ in 0..logn {
        r = (r << 1) | (k & 1);
        k >>= 1;
    }
    r
}

fn log2n(n: usize) -> usize {
    let mut logn = 0;
    let mut t = n;
    while t > 1 {
        logn += 1;
        t >>= 1;
    }
    logn
}

/// In-place forward negacyclic complex FFT over C[x]/(x^n+1).
pub fn fft(f: &mut [Complex64], n: usize) {
    let logn = log2n(n);
    let mut k = 0usize;
    let mut length = n >> 1;
    while length >= 1 {
        let mut start = 0;
        while start < n {
            k += 1;
            let brk = bit_rev(k, logn);
            let angle = PI * brk as f64 / n as f64;
            let w: Complex64 = (angle.cos(), angle.sin());
            for j in start..start + length {
                let t = cmul(w, f[j + length]);
                f[j + length] = csub(f[j], t);
                f[j] = cadd(f[j], t);
            }
            start += 2 * length;
        }
        length >>= 1;
    }
}

/// In-place inverse negacyclic complex FFT over C[x]/(x^n+1). Result scaled by 1/n.
pub fn ifft(f: &mut [Complex64], n: usize) {
    let logn = log2n(n);
    let mut k = n;
    let mut length = 1usize;
    while length < n {
        let mut start = (n - 2 * length) as isize;
        while start >= 0 {
            k -= 1;
            let brk = bit_rev(k, logn);
            let angle = -PI * brk as f64 / n as f64;
            let w_inv: Complex64 = (angle.cos(), angle.sin());
            for j in (start as usize)..(start as usize) + length {
                let t = f[j];
                f[j] = cadd(t, f[j + length]);
                f[j + length] = cmul(w_inv, csub(t, f[j + length]));
            }
            start -= (2 * length) as isize;
        }
        length <<= 1;
    }
    let inv_n = 1.0 / n as f64;
    for x in f.iter_mut() {
        x.0 *= inv_n;
        x.1 *= inv_n;
    }
}

/// Splits n-element FFT polynomial into two (n/2)-element polynomials.
/// f(x) = f0(x²) + x·f1(x²)
pub fn split_fft(f: &[Complex64], n: usize) -> (Vec<Complex64>, Vec<Complex64>) {
    let logn = log2n(n);
    let h = n / 2;
    let mut f0 = vec![(0.0f64, 0.0f64); h];
    let mut f1 = vec![(0.0f64, 0.0f64); h];
    for k in 0..h {
        let j = bit_rev(k, logn - 1);
        let angle = PI * (2 * j + 1) as f64 / n as f64;
        let omega_j: Complex64 = (angle.cos(), angle.sin());
        let a = f[2 * k];
        let b = f[2 * k + 1];
        f0[k] = cdiv_real(cadd(a, b), 2.0);
        // (a - b) / (2 * omega_j)
        let two_omega = (2.0 * omega_j.0, 2.0 * omega_j.1);
        // division by complex: (a-b) / (2*omega) = (a-b) * conj(2*omega) / |2*omega|^2
        let num = csub(a, b);
        let denom_conj = cconj(two_omega);
        let denom_sq = two_omega.0 * two_omega.0 + two_omega.1 * two_omega.1;
        f1[k] = cdiv_real(cmul(num, denom_conj), denom_sq);
    }
    (f0, f1)
}

/// Merges two (n/2)-element FFT polynomials into one n-element polynomial.
pub fn merge_fft(f0: &[Complex64], f1: &[Complex64], n: usize) -> Vec<Complex64> {
    let logn = log2n(n);
    let h = n / 2;
    let mut f = vec![(0.0f64, 0.0f64); n];
    for k in 0..h {
        let j = bit_rev(k, logn - 1);
        let angle = PI * (2 * j + 1) as f64 / n as f64;
        let omega_j: Complex64 = (angle.cos(), angle.sin());
        let t = cmul(omega_j, f1[k]);
        f[2 * k] = cadd(f0[k], t);
        f[2 * k + 1] = csub(f0[k], t);
    }
    f
}

/// Converts i32 polynomial to complex FFT domain.
pub fn int32s_to_fft(a: &[i32], n: usize) -> Vec<Complex64> {
    let mut f: Vec<Complex64> = a.iter().map(|&v| (v as f64, 0.0)).collect();
    fft(&mut f, n);
    f
}

/// Applies IFFT and rounds to nearest integer polynomial.
pub fn round_fft_to_int32s(fv: &[Complex64], n: usize) -> Vec<i32> {
    let mut tmp = fv.to_vec();
    ifft(&mut tmp, n);
    tmp.iter().map(|v| v.0.round() as i32).collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_fft_ifft_roundtrip() {
        let n = 64;
        let orig: Vec<Complex64> = (0..n).map(|i| (i as f64, 0.0)).collect();
        let mut f = orig.clone();
        fft(&mut f, n);
        ifft(&mut f, n);
        for (a, b) in orig.iter().zip(f.iter()) {
            assert!((a.0 - b.0).abs() < 1e-6, "mismatch: {} vs {}", a.0, b.0);
        }
    }

    #[test]
    fn test_fft_split_merge() {
        let n = 64;
        let orig: Vec<Complex64> = (0..n).map(|i| (i as f64 * 0.5, 0.0)).collect();
        let mut f = orig.clone();
        fft(&mut f, n);
        let (f0, f1) = split_fft(&f, n);
        let reconstructed = merge_fft(&f0, &f1, n);
        for (a, b) in f.iter().zip(reconstructed.iter()) {
            assert!((a.0 - b.0).abs() < 1e-9, "re mismatch");
            assert!((a.1 - b.1).abs() < 1e-9, "im mismatch");
        }
    }
}
